editoast/editoast_derive/src/model.rs

```diff
@@ -1,55 +1,209 @@
+
+    let create_functions = create_functions(&config);
+    let retrieve_functions = retrieve_functions(&config);
+    let delete_functions = delete_functions(&config);
 
     let expanded = quote! {
-        impl #name {
-            pub fn persist_batch(
-                values: &[Self],
-                infrastructure_id: i64,
-                conn: &mut diesel::PgConnection,
-            ) -> crate::error::Result<()> {
-                use #table::dsl::*;
-                use crate::diesel::RunQueryDsl;
-                use diesel::ExpressionMethods;
-
-                let datas = values
-                    .iter()
-                    .map(|value| {
-                        (
-                            obj_id.eq(value.get_id().clone()),
-                            data.eq(serde_json::to_value(value).unwrap()),
-                            infra_id.eq(infrastructure_id),
-                        )
-                    })
-                    .collect::<Vec<_>>();
-
-                // Work around a diesel limitation
-                // See https://github.com/diesel-rs/diesel/issues/2414
-                // Divided by 3 here because we are inserting three values
-                const DIESEL_MAX_VALUES : usize = (2_usize.pow(16) - 1)/3;
-                for data_chunk in datas.chunks(DIESEL_MAX_VALUES) {
-                    diesel::insert_into(#table::table)
-                        .values(data_chunk)
-                        .execute(conn)?;
+        #create_functions
+        #delete_functions
+        #retrieve_functions
+    };
+    Ok(expanded)
+}
+
+fn create_functions(config: &Config) -> TokenStream {
+    // If `create` is not enabled, we don't generate anything
+    if !config.create_enabled {
+        return quote! {};
+    }
+    let table = &config.table;
+    let model_name = &config.model_name;
+    let documentation = format!(
+        r#"
+        Create a new `{model_name}` in the database.
+        Returns the created object with its default values filled (like the id).
+
+        ### Example
+
+        ```
+        let obj = {model_name} {{ id: None, ... }};
+        let created_obj = obj.create(db_pool).await?;
+        let obj_id = created_obj.id.unwrap();
+        ```
+        "#
+    );
+
+    quote! {
+        #[async_trait::async_trait]
+        impl crate::models::Create for #model_name {
+            fn create_conn(self, conn: &mut diesel::PgConnection) -> crate::error::Result<Self> {
+                use #table::table;
+                use diesel::RunQueryDsl;
+
+                Ok(diesel::insert_into(table).values(&self).get_result(conn)?)
+            }
+
+            #[doc = #documentation]
+            async fn create(self, db_pool: actix_web::web::Data<crate::DbPool>) -> crate::error::Result<Self> {
+                use actix_web::web::block;
+
+                block::<_, crate::error::Result<_>>(move || {
+                    let mut conn = db_pool.get().expect("Failed to get DB connection");
```

Comment:
There are quite many `expect` like this one everywhere, wouldn't it be worth it to create a dedicated `ConnError: EditoastError` and propagate it? Because if it fails, it can be quite hard to spot where the panic originated since this is generated code.